File: src/main/scala/com/vc/medium/L1014.scala
package com.vc.medium

object L1014 {
  def maxScoreSightseeingPair(arr: Array[Int]): Int = {
    //[8, 1, 5, 2, 6]
    //res = 0
    //cur = 0

    //res = Math.max(res, arr(i) + cur)
    //cur = Math.max(cur, arr(i)) - 1

    //res = (0, 8) = 8
    //cur = (0, 8) - 1 = 7

    //res = (8, 1 + 7) = 8
    //cur = (7, 1) - 1 = 6

    //res = (8, 5 + 6) = 11
    //cur = (6, 5) - 1 = 5

    //res = (11, 2 + 5) = 11
    //cur = (5, 2) - 1 = 4

    //res = (11, 6 + 4) = 11
    //cur = (4, 6) - 1 = 5

    var i = 0
    val n = arr.length
    var currMax = 0
    var res = 0
    while(i < n) {
      res = Math.max(res, currMax + arr(i))
      currMax = Math.max(currMax, arr(i)) - 1
      i += 1
    }
    res
  }
}

File: src/main/java/com/vc/medium/L1014.java
package com.vc.medium;

class L1014 {
    public int maxScoreSightseeingPair(int[] A) {
        if(A == null || A.length == 0) return 0;

        int n = A.length, max = 0;

        int prev = max = A[0];
        for(int i = 1; i < n; i++) {
            int newPrev = Math.max(A[i], prev - 1);
            max = Math.max(max, prev - 1 + A[i]);
            prev = newPrev;
        }
        return max;
    }
}
